src/nvim/eval.c

```diff
@@ -10054,23 +10054,68 @@ static void f_getcmdwintype(typval_T *argvars, typval_T *rettv)
   rettv->vval.v_string[0] = cmdwin_type;
 }
 
-/*
- * "getcwd()" function
- */
+/// `getcwd('s')` function
+///
+/// @pre The type of the argument `s` must be a string or nothing.
 static void f_getcwd(typval_T *argvars, typval_T *rettv)
 {
-  char_u      *cwd;
+  // Pre-condition
+  if (argvars->v_type != VAR_STRING && argvars->v_type != VAR_UNKNOWN) {
+    EMSG(_(e_invarg));
+    return;
+  }
+
+  char_u *cwd = NULL;                    // current working directory to print
+  char_u *scp = argvars->vval.v_string;  // scope
 
   rettv->v_type = VAR_STRING;
   rettv->vval.v_string = NULL;
+
   cwd = xmalloc(MAXPATHL);
-  if (os_dirname(cwd, MAXPATHL) != FAIL) {
-    rettv->vval.v_string = vim_strsave(cwd);
+  cwd[0] = '\0';  // initialize to empty string
+
+  // If there is no argument return the current working directory.
+  // This is the same behaviour as in Vim.
+  //
+  // Also if global is requested and the current working directory is the
+  // global one.
+  if (argvars->v_type == VAR_UNKNOWN || (!globaldir && scp[0] == 'g')) {
+    if (os_dirname(cwd, MAXPATHL) != FAIL) {
+      goto prepare_result;
+    } else {
+      EMSG("E41: Could not allocate memory to display path.");
+      xfree(cwd);
+      return;
+    }
+  }
+
+  char_u *from = NULL;  // The original string to copy
+
+  switch (scp[0]) {  // only check the first character
+    case 'g':  // global
+      from = globaldir;
+      break;
+    case 't':  // tab
+      from = curtab->localdir;
+      break;
+    case 'w':  // window
+      from = curwin->w_localdir;
+      break;
+    default:
+      EMSG("E740: valid arguments are either 'g', 't', 'w', or nothing.");
```

Comment:
> I am completely against the tradition of writing error messages which say only that there was some problem.

My argument was only regarding consistency with the current code base, so If what you suggest is the direction we want to go in then I completely agree.